Elastic search

jakobw opened the pull request:
Passes queries with single triples of the shape `?x wdt:P123 [VALUE].` to the search API using `haswbstatement`.

- [x] Actually make requests
- [x] Literal values
- [x] Entity node values
- [x] Paginated results
- [x] Tests

Example:
```sparql
SELECT ?item WHERE {
  ?item wdt:P345 "nm0010930".
}
```
---
Update: We can now get cats from elastic too:
```sparql
SELECT ?item WHERE {
  ?item wdt:P31 wd:Q146.
}
```

jakobw committed:
Add SubjectByPropertyValueHandler

This handler forwards the query to ElasticSearch via the search API
instead of the Wikidata Query Service. It matches queries that have a
single triple of the shape `?x wdt:P123 [VALUE].`.

lucaswerkmeister commented on lib/SubjectByPropertyValueHandler.js:
```diff
@@ -0,0 +1,41 @@
+		}
+
+		return new CirrusSearchRequest(
+			triple.subject.id.substring( 1 ),
```
Maybe there should also be a check that the `SELECT` has exactly this variable and nothing else? So that we don’t match (hypothetical) queries like `SELECT ?item (NOW() AS ?asOf) WHERE { ?item wdt:P123 "456" }`

jakobw replied:
Yup good point! Done!

lucaswerkmeister commented on lib/CirrusSearchRequest.js:
```diff
@@ -0,0 +1,46 @@
+	}
+
+	respond( request, response ) {
+		nodemw.search( `haswbstatement:${this.property}=${this.value}`, ( _err, searchResults ) => {
```
This won’t find lexemes (since the Lexeme: namespace isn’t searched by default), but apparently nodemw doesn’t let you adjust the search namespaces without switching from `search` to a more generic method :/

jakobw replied:
Done! I set `srnamespace` to `*` to search all namespaces. Ideally we would probably want to search only entity namespaces, but I think this is ok for now.

lucaswerkmeister reviewed (approved):
So `nodemw`’s continuation support consists of following the continuation until it’s done, and then returning all the results at once, if I understand it correctly? I was hoping for something more streaming :/

A few comments below, but could also be resolved in follow-ups, up to you.

jakobw committed:
Use nodemw's `getAll` instead of `search`

This allows us to pass additional custom parameters to the api request.

jakobw committed:
SubjectByPropertyValueHandler: ensure subject matches result variable

Avoids accidentally matching queries like queries like
`SELECT ?item (NOW() AS ?asOf) WHERE { ?item wdt:P123 "456" }`.

jakobw committed:
CirrusSearchRequest: search all namespaces
